Utility to check whether a function is compiled with framepointers

florian-kuebler opened the pull request:
The checker checks a functions Prologue and is also looking for a matching Epilogue.
Leaf functions will be no target to that check, i.e. it is fine to have leaf functions without a correct Prologue/Epilogue.

dpallotti commented on OrbitCore/CMakeLists.txt:
```diff
@@ -5,7 +5,7 @@
 
 project(OrbitCore)
-add_library(OrbitCore STATIC)
+add_library(OrbitCore STATIC FunctionFramepointerValidator.cpp FunctionFramepointerValidator.h)
```
These don't belong here.

dpallotti commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,131 @@
+  cs_free(instructions_, instructions_count_);
+}
+
+bool FunctionFramepointerValidator::isCallInstruction(
```
Method names should be PascalCase, not camelCase.

dpallotti commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,131 @@
+  return reg == X86_REG_SP || reg == X86_REG_ESP || reg == X86_REG_RSP;
+}
+
+// we will actually only check, whether there is a correct epilogue
```
Can you rephrase better? Does it mean that the existence of one epilogue is enough?

dpallotti replied:
Pick up the habit of writing comments with a starting capital letter and a final period (per the style guide) before it's too late.

dpallotti commented on OrbitCore/FunctionFramepointerValidator.h:
```diff
@@ -0,0 +1,45 @@
+  bool validate();
+
+ private:
+  bool isCallInstruction(const cs_insn& instruction);
```
Most of these methods can be static.

dpallotti replied:
You mean free-standing functions? I would say it's up to you. They should just not be instance methods.

florian-kuebler replied:
We should discuss this offline at some later point in time. For now, I will just make them offline.

dpallotti commented on OrbitCore/FunctionFramepointerValidatorTest.cpp:
```diff
@@ -0,0 +1,50 @@
+#include "FunctionFramepointerValidator.h"
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
The copyright should be at the very top.

dpallotti commented on OrbitCore/FunctionFramepointerValidatorTest.cpp:
```diff
@@ -0,0 +1,50 @@
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
+#define WITH_FP "\x55\x89\xE5\x83\xC0\x01\xE8\x77\x00\x00\x00\x89\xEC\x5D"
```
Do these need to be `define`s? Can they just be constant?

florian-kuebler replied:
Thanks for spotting this. Please see above.

dimitry- commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,131 @@
+bool FunctionFramepointerValidator::validateEpilogue() {
+  bool found_epilogue = false;
+  for (size_t i = 0; i < instructions_count_ - 1; i++) {
+    if (isMovInstruction(instructions_[i]) &&
```
do you check if there is ret after mov and pop?

florian-kuebler replied:
In capstone there is: 
```// all return instructions
X86_GRP_RET```

So it should be safe to use this group, as done for calls.

But I don't get you comment with pop rip. Could you elaborate?

dimitry- replied:
if you have return address on the top of the stack pop rip is effectively a ret. The question is if amd64 lets regular instructions modify rip (or directly assign something to rip), for example arm32 lets you do that.

From what I understand amd64 does not allow that so we should be fine with checking RET

florian-kuebler replied:
Ah thanks for clarifying. Well lets implement the check for RET, we would only produce false positive warnings if pop rip would be possible though.

dpallotti commented on OrbitCore/FunctionFramepointerValidator.h:
```diff
@@ -0,0 +1,45 @@
+
+#include <capstone/capstone.h>
+
+/*
```
We are not really using this style for class comments. Use `//` on every line.

dpallotti commented on OrbitCore/FunctionFramepointerValidator.h:
```diff
@@ -0,0 +1,45 @@
+                                size_t code_size);
+  virtual ~FunctionFramepointerValidator();
+
+  // FunctionFramepointerValidator is neither copyable nor movable.
```
Then delete the move constructor and assignment operator as well.

florian-kuebler replied:
Okay, Done.

dimitry- commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,131 @@
+  return true;
+}
+
+bool FunctionFramepointerValidator::validate() {
```
how about exception for _start which does some special things?

dimitry- replied:
Actually never mind, this method probably is not called at all for _start

dimitry- replied:
yes. And also we probably do not even care what it is called because the way to add expection would be to examine entry_point in the Elf_Ehdr

dpallotti commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,131 @@
+      uint8_t read_count_first, write_count_first, read_count_second,
+          write_count_second;
+      // check for "mov esp, ebp; pop ebp"
+      if (  // first instruction
```
This condition is unreadable. I would split it in a few boolean variables, *at least* one for "first instruction" and one for "second instruction", but possibly more.

dpallotti commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,131 @@
+}
+
+bool FunctionFramepointerValidator::validateFramePointers() {
+  return validateEpilogue() && validatePrologue();
```
Isn't it weird to have `validatePrologue` second?

dpallotti replied:
Same goes for the order of declaration and definition in the class.

dpallotti commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,131 @@
+}
+
+bool FunctionFramepointerValidator::validate() {
+  if (instructions_count_ > 0) {
```
What about an early exit?
```
if (instruction_count_ == 0) {
  ...
  return false;
}

<unindented code>
```

florian-kuebler commented on OrbitCore/FunctionFramepointerValidatorTest.cpp:
```diff
@@ -0,0 +1,50 @@
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
+#define WITH_FP "\x55\x89\xE5\x83\xC0\x01\xE8\x77\x00\x00\x00\x89\xEC\x5D"
```
Sure, thanks, I adjusted them accordingly. The defines were basically from the developer guide of capstone.

dimitry- commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -19,113 +19,119 @@ FunctionFramepointerValidator::~FunctionFramepointerValidator() {
+// It might be the case, that there are multiple function returns and that not
+// all are correct. However, we would not expect a compiler to produce this,
+// and for hand written assembly, we accept wrong unwinding results.
+bool FunctionFramepointerValidator::ValidateEpilogue() {
```
to make it more readable use early returns..

florian-kuebler replied:
Okay, sure.

dimitry- commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -19,113 +19,119 @@ FunctionFramepointerValidator::~FunctionFramepointerValidator() {
+bool FunctionFramepointerValidator::ValidateEpilogue() {
+  bool found_epilogue = false;
+  for (size_t i = 0; i < instructions_count_ - 1; i++) {
+    if (IsMovInstruction(instructions_[i]) &&
```
here you can do
if (!IsMov ... ) {
  continue;
}

florian-kuebler replied:
Good point

dimitry- commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -19,113 +19,119 @@ FunctionFramepointerValidator::~FunctionFramepointerValidator() {
+          cs_regs_access(handle_, &instructions_[i], regs_read_first,
+                         &read_count_first, regs_write_first,
+                         &write_count_first) == 0;
+
```
if succeeded_regs_access_first is false you can probably continue... since in this case checks below do not make sense, right?

and so on..

florian-kuebler replied:
Thanks for spotting this. I made them if(...) continues.

dimitry- commented on OrbitCore/FunctionFramepointerValidatorTest.cpp:
```diff
@@ -0,0 +1,57 @@
+
+#define NO_FP \
+  "\x29\x25\x00\x00\x00\x00\xE8\xFD\xFF\xFF\xFF\x01\x25\x00\x00\x00\x00\xC3"
+#define LEAF "\x29\x24\x25\x00\x00\x00\x00\x48\x01\x24\x25\x00\x00\x00\x00\xC3"
```
forgot to remove these?

florian-kuebler replied:
Got me. Thanks

dpallotti commented on OrbitCore/FunctionFramepointerValidator.h:
```diff
@@ -0,0 +1,46 @@
+      const FunctionFramepointerValidator&) = delete;
+  FunctionFramepointerValidator(FunctionFramepointerValidator&&) = delete;
+  FunctionFramepointerValidator& operator=(
+      const FunctionFramepointerValidator&&) = delete;
```
Remove this const as it's an rvalue-ref.

florian-kuebler replied:
Okay. Thanks.

dpallotti commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,137 @@
+
+// We will actually only check, whether there is any correct epilogue.
+// It might be the case, that there are multiple function returns and that not
+// all are correct. However, we would not expect a compiler to produce this,
```
I'm pretty sure compilers produce multiple ret instructions in optimized code (maybe not together with "mov esp, ebp" and "pop ebp", but for the sake of this comment...)

dpallotti commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,137 @@
+          IsStackPointer(regs_write_second[0]) &&
+          IsBasePointer(regs_write_second[1]);
+
+      // Are both instructions as expected? If so, we found an epilogue
```
What about the "leave" instruction, which is equivalent to "mov esp, ebp","pop ebp"?

florian-kuebler replied:
Good point. I wasn't aware of this instruction. I added an early return if we find this instruction.

dimitry- commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,155 @@
+    cs_regs regs_read_first, regs_write_first, regs_read_second,
+        regs_write_second;
+    uint8_t read_count_first, write_count_first, read_count_second,
+        write_count_second;
```
naming here is still a bit confusing.

maybe remove first and second and reuse them.
regs_read = source_registers
regs_write = destination registers

dimitry- replied:
On other hand for pop instruction it is not even that... maybe use different naming there

florian-kuebler replied:
I would be fine with just having `regs_read` and `regs_write`

florian-kuebler commented:
I updated the PR with handling of the RET instructions as well as simplification of the naming.
Further, I added handling of ENTER/LEAVE instructions.

dimitry- commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -47,6 +55,10 @@ bool FunctionFramepointerValidator::ValidatePrologue() {
   // check the first instruction: it must be "push ebp"
```
or enter apparently...

florian-kuebler replied:
Thanks.

dimitry- commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -47,6 +55,10 @@ bool FunctionFramepointerValidator::ValidatePrologue() {
                      regs_write, &write_count) == 0) {
+    if (instructions_[0].id == X86_INS_ENTER) {
+      return true;
+    }
```
ok, this probably introduces another state... the question here is, is it valid to use ENTER without LEAVE and LEAVE without ENTER.

florian-kuebler replied:
As discussed offline, this would be strange hand-crafted assembly, which would silently accept here.

dimitry- commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -71,9 +83,10 @@ bool FunctionFramepointerValidator::ValidatePrologue() {
 bool FunctionFramepointerValidator::ValidateEpilogue() {
-  for (size_t i = 0; i < instructions_count_ - 1; i++) {
+  for (size_t i = 0; i < instructions_count_ - 2; i++) {
     // "leave" is equivalent to "mov esp, ebp" "pop ebp"
```
not always - apparently it depends on nesting attribute. it could do more if I understand correctly

florian-kuebler replied:
As discussed offline, we would still accept this here.

dimitry- commented on OrbitCore/FunctionFramepointerValidatorTest.cpp:
```diff
@@ -11,7 +11,7 @@ namespace {
+                                       0x00, 0x89, 0xEC, 0x5D, 0xC3};
 constexpr uint8_t kFunctionWithoutFP[] = {0x29, 0x25, 0x00, 0x00, 0x00, 0x00,
                                           0xE8, 0xFD, 0xFF, 0xFF, 0xFF, 0x01,
                                           0x25, 0x00, 0x00, 0x00, 0x00, 0xC3};
```
do you want to add test for ENTER and LEAVE - including invalid ones maybe, depending on what we decide.

florian-kuebler replied:
Added a test for this and also for tail calls. Thanks.

dimitry- commented on OrbitCore/FunctionFramepointerValidator.cpp:
```diff
@@ -0,0 +1,172 @@
+// epilogue is the return to the caller.
+// TODO(kuebler): Better handling for tail call optimization
+bool FunctionFramepointerValidator::ValidateEpilogue() {
+  for (size_t i = 0; i < instructions_count_ - 2; i++) {
```
- 2 here could be invalid in the case when last 2 instructions are LEAVE and RET, right? Maybe change it to while true and introduce instruction iterator to avoid such corner cases here.

florian-kuebler replied:
Thanks for spotting. I fixed this accordingly.
